contracts/validators/CredentialAtomicQuerySigValidatorV2.sol

```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "../lib/GenesisUtils.sol";
+import "../interfaces/ICircuitValidator.sol";
+import "../interfaces/IVerifier.sol";
+import "../interfaces/IState.sol";
+
+contract CredentialAtomicQuerySigValidatorV2 is
+    OwnableUpgradeable,
+    ICircuitValidator
+{
+    string constant CIRCUIT_ID = "credentialAtomicQuerySig";
+    uint256 constant CHALLENGE_INDEX = 4;
+    uint256 constant USER_ID_INDEX = 2;
+
+    IVerifier public verifier;
+    IState public state;
+
+    uint256 public revocationStateExpirationTime;
+
+    function initialize(
+        address _verifierContractAddr,
+        address _stateContractAddr
+    ) public initializer {
+        revocationStateExpirationTime = 1 hours;
+        verifier = IVerifier(_verifierContractAddr);
+        state = IState(_stateContractAddr);
+        __Ownable_init();
+    }
+
+    function setRevocationStateExpirationTime(uint256 expirationTime)
+        public
+        onlyOwner
+    {
+        revocationStateExpirationTime = expirationTime;
+    }
+
+    function getCircuitId() external pure returns (string memory id) {
+        return CIRCUIT_ID;
+    }
+
+    function getChallengeInputIndex() external pure returns (uint256 index) {
+        return CHALLENGE_INDEX;
+    }
+
+    function getUserIdInputIndex() external pure returns (uint256 index) {
+        return USER_ID_INDEX;
+    }
+
+    function verify(
+        uint256[] memory inputs,
+        uint256[2] memory a,
+        uint256[2][2] memory b,
+        uint256[2] memory c,
+        CircuitQuery memory query
+    ) external view returns (bool r) {
+        // verify that zkp is valid
+        require(
+            verifier.verifyProof(a, b, c, inputs),
+            "atomic query signature proof is not valid"
+        );
+
+        // verify query
+        require(
+            inputs[8] == query.schema,
+            "wrong claim schema has been used for proof generation"
+        );
+        require(
+            inputs[9] == query.slotIndex,
+            "wrong claim data slot has been used for proof generation"
+        );
+        require(
+            inputs[10] == query.operator,
+            "wrong query operator has been used for proof generation"
+        );
+        require(
+            inputs[1] == query.valueHash,
+            "wrong comparison value has been used for proof generation"
+        );
+
+        // verify user states
+
+        uint256 userId = inputs[USER_ID_INDEX];
+        uint256 userState = inputs[3];
+        uint256 issuerAuthState = inputs[0];
+        uint256 issuerId = inputs[5];
+        uint256 issuerClaimNonRevState = inputs[6];
+
+        // 1. User state must be latest or genesis
```

Comment:
May be better move `User state must be latest or genesis` check to utils? We're repeating it a lot

Reply:
Is this check will always be the same between Mtp and Sig contracts?